pubtools/pulplib/_impl/fake/client.py

```diff
@@ -79,6 +82,35 @@ def get_repository(self, repository_id):
 
         return f_return(data[0])
 
+    def _do_upload_file(self, upload_id, file_obj, name):
+        # pylint: disable=unused-argument
+        out = f_return((hashlib.sha256().hexdigest(), random.randint(10, 1000)))
+
+        if "close" not in dir(file_obj):
+            file_obj = open(file_obj, "rb")
+            out.add_done_callback(lambda _: file_obj.close())
+
+        return out
+
+    def _request_upload(self):
+        upload_request = {
+            "_href": "/pulp/api/v2/content/uploads/%s/" % self._request_id(),
+            "upload_id": "%s" % self._request_id,
+        }
+
+        return f_return(upload_request)
+
+    def _do_import(self, repo_id, upload_id, unit_type_id, unit_key):
+        # pylint: disable=unused-argument
+        task = Task(id=self._next_task_id(), completed=True, succeeded=True)
+
+        self._upload_history.append(Upload(repo_id, [task]))
```

Comment:
Can you please make this fail if the repo doesn't exist, like a real client would?

Reply:
Is this really necessary?
The use case for upload is to get repository first and then call repo.upload_file. If the repo doesn't exist, it should fail at get_repository, do we need to double-check here?

Reply:
I got this reply from Jay but now I can't find it in github for some reason:

> Is this really necessary?
> The use case for upload is to get repository first and then call repo.upload_file.
> If the repo doesn't exist, it should fail at get_repository, do we
> need to double-check here?

I would say that it is indeed still necessary since there can be any number of references to
the same repository.  I think there's already some autotests on the publish side which cover this.

For example you can have:

```
repo1 = client.get_repository('some-repo').result()
repo2 = client.get_repository('some-repo').result()

repo1.delete().result()

# and now try to upload to repo2
# expected behavior: it should fail!
repo2.upload_file(...)
```